src/async_interface.rs

```diff
@@ -0,0 +1,112 @@
+
+use std::collections::VecDeque;
+use std::pin::Pin;
+
+use libc::c_void;
+use libusb1_sys::libusb_transfer;
+
+use crate::Context;
+use crate::DeviceHandle;
+
+#[derive(Debug)]
+pub enum TransferStatus {
+	Completed(Vec<u8>),
+	Error,
+	TimedOut,
+	Stall,
+	NoDevice,
+	Overflow,
+	Unknown(i32)
+}
+
+// Putting the libusb_transfer pointer and the buffer in the struct that owns both ties the lifetimes
+// together and ensures that `buff` will live as long as `ptr`.  We also know that the receiving VecDeque
+// will live as long as the pointer and buffer because it's owned by the same struct.  We put the pointer
+// inside `recv` to the `libusb_transfer` struct, so it's important for `recv` to stay at the same address,
+// which is why it's wrapped in a Pin<_>
+// This is not thread-safe at all.  If you try to implement Send or Sync on it, it'll have to be an unsafe impl.
+pub struct AsyncTransfer {
+	pub ptr: *mut libusb_transfer,
+	pub buff: Vec<u8>,
+	pub recv: Pin<Box<VecDeque<TransferStatus>>>,
```

Comment:
I'm not sure `AsyncTransfer` or specifically `recv` is prevented from moving here. Neither the VecDeque or `TransferStatus` is `!Unpin`, so neither is `AsyncTransfer`. If this is true, it means that when an `AsyncTransfer` is moved, we get a dangling reference inside of `ptr`.

BUT I'm very unfamiliar with unsafe rust, or self-referential rust. Am I correct in this? Just going off of the `std::pin` documentation.